### PR TITLE
Login page firestore

### PR DESCRIPTION
**Pull request by suivo2001**
Added login page with css, firestore structure and rules almost finalized

### CODE REVIEW

File: firestore.rules [hunk 1]
```diff
@@ -1,11 +1,11 @@
...
-
-    
-    match /{document=**} {
-      allow read: if false;
-      allow write: if true;
+    match /participant_responses/{study_id}/participants/{participant_id}/data/{document=**} {
+          allow read: if false
+          allow create, update: 
+          if participant_id in get(/databases/$(database)/documents/registered_studies/$(study_id)).data.registered_participants &&
+              exists(/databases/$(database)/documents/registered_studies/$(study_id));
```
Comment (mcmcgrath13) [inline firestore.rules:8]: is this second one needed? if the first `get` works, mustn't the `second`?

File: package.json [hunk 2]
```diff
@@ -27,7 +27,9 @@
     "lodash": "^4.17.11",
     "popper.js": "^1.16.0",
     "react": "^16.8.6",
+    "react-bootstrap": "^1.6.0",
```
Comment (mcmcgrath13) [inline package.json:30]: do we need react-bootstrap? bootstrap is already included as a dep?

File: src/App.js [hunk 3]
```diff
@@ -76,4 +72,21 @@ class App extends React.Component {
...
 
+class App extends React.Component {
+  render() {
+    return (
+      <>
+        <Router>
+          <Switch>
+          <Route exact path="/">
+            {LOGGEDIN ? <Redirect to="/experiment" /> : <Login/>}
+          </Route>
```
Comment (mcmcgrath13) [inline src/App.js:83]: route feels like a heavy construct for this? what about just an if around whether or not logged in?

File: src/firebase.js [hunk 4]
```diff
@@ -26,35 +60,63 @@ if (window.location.hostname === "localhost") {
...
+      return false
+    });
 };
 
-// create a document in the collection with a random id
-const createFirebaseDocumentRandom = () => {
-  db.collection(collectionName).add({ dateCreated: new Date() });
-};
+/*
+const addToEnd = (patientId, studyId, data) =>{
```
Comment (mcmcgrath13) [inline src/firebase.js:84]: remove commented out code

File: src/firebase.js [hunk 5]
```diff
@@ -26,35 +60,63 @@ if (window.location.hostname === "localhost") {
...
+    .collection('participants')
     .doc(patientId)
-    .collection("data")
-    .doc(`trial_${data.trial_index}`)
-    .set(data);
+    .collection('data')
+    .doc(timestamp)
+    .update('results', firebase.firestore.FieldValue.arrayUnion(data))
+    //.collection(date)
+    //.doc(`trial_${data.trial_index}`)
```
Comment (mcmcgrath13) [inline src/firebase.js:110]: remove commented out code

File: src/lib/utils.js [hunk 6]
```diff
@@ -78,7 +78,14 @@ const getTurkUniqueId = () => {
 const getUserId = (data) => {
   const patientId = JSON.parse(data.responses)['Q0']
   jsPsych.data.addProperties({patient_id: patientId, timestamp: Date.now()})
-  console.log("ID", patientId)
+  console.log("PatientID", patientId)
+}
```
Comment (mcmcgrath13) [inline src/lib/utils.js:82]: are these functions still being used?

File: src/login.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+import Form from "react-bootstrap/Form";
+import Button from "react-bootstrap/Button";
+import {createPatient} from './firebase'
+import { useHistory } from 'react-router-dom';
+import { jsPsych } from 'jspsych-react'
+
+ function Login() {
+    const dateTimestamp = Date.now()
+    const curDate = new Date(dateTimestamp)
+    const [patientId, setPatient] = useState("");
```
Comment (mcmcgrath13) [inline src/login.js:11]: run `prettier` on this file to get nice formatting

File: src/timelines/preamble.js [hunk 8]
```diff
@@ -18,7 +18,7 @@ else if (AT_HOME && VIDEO) {
   tl = [experimentStart(), userId(), instructions(), camera()]
 }
 else {
-  tl = [experimentStart(), userId(), adjustVolume(), holdUpMarker(), startCode(), instructions()]
+  tl = [experimentStart(), /*userId(), */adjustVolume(), holdUpMarker(), startCode(), instructions()]
```
Comment (mcmcgrath13) [inline src/timelines/preamble.js:21]: remove commented out code

File: src/trials/userId.js [hunk 9]
```diff
@@ -1,7 +1,10 @@
...
-import { getUserId, getTurkUniqueId } from '../lib/utils'
+import { getUserId, getStudyId, getTurkUniqueId } from '../lib/utils'
 import { baseStimulus } from '../lib/markup/stimuli'
 
+
+
+
 const userId = () => {
   if (MTURK) {
     return {
```
Comment (mcmcgrath13) [inline src/trials/userId.js:10]: is this file still needed?

**Review by mcmcgrath13 — CHANGES_REQUESTED** [review]

File: src/App.js (hunk 3 above)
Comment (mirestrepo) [reply src/App.js:83]: Totally agree, I think it should just be a switch between `<Login/>` and `<ExpStart/>`

File: src/App.js [hunk 10]
```diff
@@ -76,4 +72,21 @@ class App extends React.Component {
...
+            <Route path='/experiment' component={ExpStart} />
```
Comment (mirestrepo) [inline src/App.js:84]: Could we also make `ExpStart` a component in its own file? And maybe even move it together with `login.js` to a `components` folder?

File: src/login.js [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    const history = useHistory();
+
+    function validateForm() {
+        return patientId.length > 0 && studyId.length > 0;
+      }
+
+    function handleSubmit(e) {
+        e.preventDefault();
+        const logPatient = async() =>{
+            const loggedIn = await createPatient(patientId, studyId, startDate, timestamp)
```
Comment (mirestrepo) [inline src/login.js:24]: I think `loggedIn` should come form `useState` as well and remove the `LOGGEDIN`variable in `App.js`

File: src/login.js [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+            
+            if(loggedIn){
+                jsPsych.data.addProperties({patient_id: patientId, study_id: studyId, timestamp: timestamp, start_date: startDate})
+                history.push('/experiment')
```
Comment (mirestrepo) [inline src/login.js:28]: If made a component, this won't be necessary

**Review by mirestrepo — CHANGES_REQUESTED** [review]

File: src/login.js (hunk 12 above)
Comment (mcmcgrath13) [reply src/login.js:28]: I'm confused - isn't it already a component? Instead this would need to pass the action to set logged in on the parent component, no?

File: src/App.js (hunk 10 above)
Comment (broarr) [reply src/App.js:84]: Technically, that has to do with the transpiler you use, but it is good practice

File: src/firebase.js [hunk 13]
```diff
@@ -3,6 +3,40 @@ import 'firebase/firestore';
...
+} else if (sUsrAg.indexOf("Safari") > -1) {
+  sBrowser = "Apple Safari";
+  // "Mozilla/5.0 (iPhone; CPU iPhone OS 11_4 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/11.0 Mobile/15E148 Safari/604.1 980x1306"
+} else {
+  sBrowser = "unknown";
+}
+
+return sBrowser
+}
+
```
Comment (broarr) [inline src/firebase.js:39]: Why not just save the user agent string instead of trying to parse it?

File: src/login.js (hunk 12 above)
Comment (broarr) [reply src/login.js:28]: History is part of the `react-router` library. I think @suivo2001 is trying to preserve the browser back button behavior here. I think that's a good reason for a router, but in this case I still don't think it matters that much

Comment (mcmcgrath13) [reply src/login.js:28]: interesting - I think that's actually a really good reason to not use `router` in this case - participants shouldn't be able to go back.  We'd also need to protect `/experiment` so that one can't navigate to there without logging in first.

File: firestore.rules (hunk 1 above)
Comment (suivo2001) [reply firestore.rules:8]: Yes, it wasn't needed and I deleted the line.

File: src/login.js (hunk 12 above)
Comment (mirestrepo) [reply src/login.js:28]: ... I meant if we render `<ExpStart>` directly component instead of routing to `/experiment`... I think we are all thinking of the same and I thing you make a good point about not being able to go back in the history nor navigate directly to the `/experiment` route

File: package.json (hunk 2 above)
Comment (suivo2001) [reply package.json:30]: The button and form I used in the login were from this package, and I think it's separate.

File: src/lib/utils.js (hunk 6 above)
Comment (suivo2001) [reply src/lib/utils.js:82]: Yeah those are not really used anymore and I deleted them.

File: src/trials/userId.js (hunk 9 above)
Comment (suivo2001) [reply src/trials/userId.js:10]: The file should not be needed anymore but it still contains some MTURK logic, so I'll delete it once me and Galen get the logic down.

File: src/firebase.js (hunk 13 above)
Comment (suivo2001) [reply src/firebase.js:39]: In that case, we probably don't need navigator.userAgent then since navigator.version which I already have included provides pretty much the same information.

File: src/App.js [hunk 14]
```diff
@@ -1,79 +1,43 @@
...
 
+  const [ipcRenderer, setRenderer] = useState(false);
+  const [psiturk, setPsiturk] = useState(false);
 
+  // All Mturk and other login logic will go here
+  useEffect(() => {
+    if (IS_ELECTRON) {
+      const electron = window.require("electron");
+      setRenderer(electron.ipcRenderer);
+    } else if (MTURK) {
```
Comment (mcmcgrath13) [inline src/App.js:24]: the old login trial also (I believe) had a way to get login info from the environment if electron - could you check into that and add here as well?

File: src/components/Login.js [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+        jsPsych.data.addProperties({
+          patient_id: patientId,
+          study_id: studyId,
+          timestamp: timestamp,
+          start_date: startDate,
+        });
+        login();
+      }
+    };
+    logPatient();
```
Comment (mcmcgrath13) [inline src/components/Login.js:51]: would it be cleaner to do this with callbacks than with the nested async function? @broarr @mirestrepo

File: src/components/Login.js [hunk 16]
```diff
@@ -0,0 +1,90 @@
+import React, { useState, useContext } from "react";
+import Form from "react-bootstrap/Form";
+import Button from "react-bootstrap/Button";
+import { createPatient } from "../firebase";
+import { jsPsych } from "jspsych-react";
+import "../App.css";
+import "bootstrap/dist/css/bootstrap.css";
+import "@fortawesome/fontawesome-free/css/all.css";
+import LoginContext from "../contexts/LoginContext";
```
Comment (mcmcgrath13) [inline src/components/Login.js:9]: @broarr is a context the right way to handle the passing back up of state? or could an action be passed as a prop? (not sure what the current react-idiomatic way is)

File: src/config/main.js [hunk 17]
```diff
@@ -29,10 +29,20 @@ const audioCodes = {
 const MTURK = (!jsPsych.turk.turkInfo().outsideTurk)
 const AT_HOME = (process.env.REACT_APP_AT_HOME === 'true')
 const VIDEO =  (process.env.REACT_APP_VIDEO === 'true')
+const ONLINE = (process.env.REACT_APP_ONLINE=== 'true')
```
Comment (mcmcgrath13) [inline src/config/main.js:32]: `ONLINE` doesn't seem to be used anywhere - is it needed?

Comment (suivo2001) [reply src/config/main.js:32]: I think this relates more to what Galen is doing and I'm not sure if it is necessary yet, so I think we could keep it for now.

File: src/App.js (hunk 14 above)
Comment (suivo2001) [reply src/App.js:24]: I think its this line `const envPatientId = ipcRenderer.sendSync('syncPatientId')`. I'm not entirely sure what this sendSync does and it doesn't work if I just directly copy the code over.

Comment (suivo2001) [reply src/App.js:24]: @mcmcgrath13 Ok I got it to work now by editing the electron.js file. It should autofil the login fields if the environment variables for patientId and studyId were set.

File: src/firebase.js [hunk 18]
```diff
@@ -26,35 +26,48 @@ if (window.location.hostname === "localhost") {
...
-};
-
-// create a document in the collection with a random id
-const createFirebaseDocumentRandom = () => {
-  db.collection(collectionName).add({ dateCreated: new Date() });
+    .collection('data')
+    .doc(timestamp)
+    .set({start_time: startDate, app_version: window.navigator.appVersion, app_platform: window.navigator.platform, results: []})
+    .then(()=>{
+      console.log('success')
```
Comment (mirestrepo) [inline src/firebase.js:40]: Remove the `console.log`?

File: src/components/Login.js (hunk 15 above)
Comment (mirestrepo) [reply src/components/Login.js:66]: I don't have strong feelings about the async/await here, but I'm not an expert on async.... I'm wondering though if `logPatient` could live in the `firebase.json` file were `createPatient` is?

Comment (mirestrepo) [reply src/components/Login.js:66]: Also I think we decided to use Participant instead of Patient. Right?

Comment (mirestrepo) [reply src/components/Login.js:66]: And one last question is what happens if the login screen is shown when `ONLINE=false` and the participant ID is not passed as an `ENV`? The code here for `logPatient` is firebase specific, so do we need to make sure we check for that?

File: src/App.js [hunk 19]
```diff
@@ -29,14 +26,19 @@ function App() {
...
+  }
+
   return (
-    <LoginContext.Provider value={{ loggedIn, login }}>
+    <>
       {loggedIn ? (
         <JsPsychExperiment ipcRenderer={ipcRenderer} psiturk={psiturk} />
       ) : (
-        <Login ipcRenderer={ipcRenderer}/>
+        <Login ipcRenderer={ipcRenderer} callBack = {handleCallBack}/>
```
Comment (mcmcgrath13) [inline src/App.js:39]: instead of `callBack` maybe `onLogin` and then just pass `setLogin`.

File: src/components/Login.js [hunk 20]
```diff
@@ -60,7 +56,7 @@ function Login({ipcRenderer}) {
           timestamp: timestamp,
           start_date: startDate,
         });
-        login();
+        callBack();
```
Comment (mcmcgrath13) [inline src/components/Login.js:59]: then change here to `onLogin(loggedIn)`

File: src/components/Login.js (hunk 15 above)
Comment (mcmcgrath13) [reply src/components/Login.js:62]: good point, there should be a check for FIREBASE

File: public/electron.js [hunk 21]
```diff
@@ -176,6 +177,11 @@ ipc.on('syncPatientId', (event) => {
   event.returnValue = process.env.REACT_APP_PATIENT_ID
 })
 
+// Get Study Id from environment
```
Comment (mcmcgrath13) [inline public/electron.js:180]: should the syncing of PatientID and study ID be two different events? or should/could we do this in one go.  I also wonder since these aren't actually build time react env vars, if we should switch up the naming? maybe `BEADS_TASK_PATIENT_ID` or more generically to be used across many tasks `TASK_APP_PATIENT_ID` (and `STUDY_ID` respectively) @mirestrepo what do you think?

File: src/App.js (hunk 14 above)
Comment (mcmcgrath13) [reply src/App.js:21]: where are you sending the sync to get the env vars?

File: src/components/Login.js (hunk 15 above)
Comment (mcmcgrath13) [reply src/components/Login.js:62]: also, need the logic to get the mturk id still I think (unless I missed that)

Comment (mcmcgrath13) [reply src/components/Login.js:62]: I think we need the jspsych updates here (as that's global to all use cases), but need more ways to set loggedIn (and think about how to make that cleanest/most re-usable/abstract

Comment (suivo2001) [reply src/components/Login.js:62]: @mirestrepo @mcmcgrath13 How would checking for FIREBASE look like? If FIREBASE == false, do we just let them enter any credentials in the login screen to continue? 
The logic for MTURK is in the useEffect in app and I just added setLogin(true) when MTURK. 
What I had in mind is for the logic for other login to be in useEffect in App.js and login.js to be the login screen, which is pretty much tied to FIREBASE.

File: src/App.js (hunk 14 above)
Comment (suivo2001) [reply src/App.js:21]: That is in login.js since I'm just using it to autofill the fields.

File: public/electron.js (hunk 21 above)
Comment (mirestrepo) [reply public/electron.js:180]: I'ma bit confused here as to what is the expectation for Electron? Are we now requiring always `PARTICIPANT_ID` and `STUDY_ID`?  In firebase we are requiring both as all data lives in the same database. I don't know the organization in the file-system, but in the "clinic" case, I feel like the scenario of having the STUDY_ID come from and ENV and the PARTICIPANT_ID be input at run time is a possibility?

`syncCredentials()` sounds reasonable to me, and naming the variables differently as well. What about just `PARTICIPANT_ID` and `STUDY_ID`? Any good reason for a prefix?

File: src/components/Login.js (hunk 15 above)
Comment (mirestrepo) [reply src/components/Login.js:62]: In my mind the login for Firebase does a credentials check, the login for Electron just sets variables - unless they were set by the ENVS..

Comment (mirestrepo) [reply src/components/Login.js:62]: Kind of the `if FIREBASE` is just around the `createPatient` call or something like that?

File: public/electron.js [hunk 22]
```diff
@@ -302,7 +298,7 @@ app.on('will-quit', () => {
     const name = app.getName()
     const today = new Date(Date.now())
     const date = today.toISOString().slice(0,10)
-    const copyPath = path.join(desktop, dataDir, `${patientID}`, date, name)
+    const copyPath = path.join(desktop, dataDir, `${participantID}`, date, name)
```
Comment (mcmcgrath13) [inline public/electron.js:301]: should this include `studyId` now too?

File: public/electron.js [hunk 23]
```diff
@@ -164,32 +164,28 @@ ipc.on('trigger', (event, args) => {
...
 
 // Read version file (git sha and branch)
 var git = JSON.parse(fs.readFileSync(path.resolve(__dirname, 'config/version.json')));
 
-// Get Patient Id from environment
-ipc.on('syncPatientId', (event) => {
-  event.returnValue = process.env.REACT_APP_PATIENT_ID
+// Get Participant Id and Study Id from environment
+ipc.on('syncCredentials', (event) => {
+  event.returnValue = [process.env.REACT_APP_PARTICIPANT_ID, process.env.REACT_APP_STUDY_ID]
```
Comment (mcmcgrath13) [inline public/electron.js:177]: would json be easier here than an array?

File: public/electron.js [hunk 24]
```diff
@@ -164,32 +164,28 @@ ipc.on('trigger', (event, args) => {
 let stream = false
 let fileName = ''
 let filePath = ''
-let patientID = ''
+let participantID = ''
```
Comment (mcmcgrath13) [inline public/electron.js:167]: do we need `studyID` too?

File: src/App.js [hunk 25]
```diff
@@ -4,39 +4,78 @@ import "bootstrap/dist/css/bootstrap.css";
...
+      }
+      setValidation('desktop')
+    // If online
+    } else {
+      // If MTURK
+      if (MTURK) {
+        /* eslint-disable */
+        window.lodash = _.noConflict();
+        setPsiturk(new PsiTurk(getTurkUniqueId(), "/complete"));
+        setLogin(true)
```
Comment (mcmcgrath13) [inline src/App.js:54]: don't we need to get the participant ID from psyTurk and save that to participant ID as well? what about the study ID here?

File: src/components/Login.js [hunk 26]
```diff
@@ -1,83 +1,87 @@
...
 import Button from "react-bootstrap/Button";
-import { createPatient } from "../firebase";
+import {  validateParticipant } from "../firebase";
 import { jsPsych } from "jspsych-react";
 import "../App.css";
 import "bootstrap/dist/css/bootstrap.css";
 import "@fortawesome/fontawesome-free/css/all.css";
 
-function Login({ipcRenderer, callBack}) {
+function Login({onLogin, participant_id, study_id, validation}) {
```
Comment (mcmcgrath13) [inline src/components/Login.js:10]: would call the props something more distinct than the entered values - the nuance between the snake and camel case versions will be easy to miss (maybe `envParticipantId` or something)

File: src/App.js [hunk 27]
```diff
@@ -4,39 +4,78 @@ import "bootstrap/dist/css/bootstrap.css";
...
+      }
+    }
+    
+  },[]);
 
   return (
     <>
       {loggedIn ? (
         <JsPsychExperiment ipcRenderer={ipcRenderer} psiturk={psiturk} />
       ) : (
```
Comment (mcmcgrath13) [inline src/App.js:77]: would this then show the Login screen even in the mturk and prolific cases? If we have that info from those sources, we should show the screen at all.  Move the jspsych adding patientId and studyId here.  Maybe pass up those from the login screen instead of the loggedIn state.  Maybe also pass an actual validation function instead of a string, where the validation function takes a study id and participant id and returns true or false.  This will keep the login component pretty slim, and keep the logic mostly here.

File: src/firebase.js [hunk 28]
```diff
@@ -26,13 +26,13 @@ if (window.location.hostname === "localhost") {
   db.useEmulator("localhost", 8080);
 }
 
-// Add patient data and trial data to db
-const createPatient = (patientId, studyId, startDate, timestamp) => {
-  // return promise with value true if patient and study id match, false otherwise
+// Add participant data and trial data to db
+const validateParticipant = (participantId, studyId, startDate, timestamp) => {
```
Comment (mcmcgrath13) [inline src/firebase.js:30]: this does more than validate, I would keep it as `createParticipant` or `initParticipant`

File: public/electron.js (hunk 24 above)
Comment (broarr) [reply public/electron.js:167]: It's right below

File: src/App.js [hunk 29]
```diff
@@ -1,79 +1,104 @@
...
+import JsPsychExperiment from "./components/JsPsychExperiment";
+import { MTURK, IS_ELECTRON, AT_HOME, FIREBASE, PROLIFIC } from "./config/main";
+import { getTurkUniqueId } from "./lib/utils";
 
+function App() {
+  // Variables for login
+  const [loggedIn, setLogin] = useState(false);
+  const [ipcRenderer, setRenderer] = useState(false);
+  const [psiturk, setPsiturk] = useState(false);
+  const [envParticipantId, setEnvParticipantId] = useState("");
```
Comment (broarr) [inline src/App.js:16]: I don't know that this needs to be state either. You can make a `let` variable when this component is first created

File: src/App.js [hunk 30]
```diff
@@ -1,79 +1,104 @@
...
+  const [envStudyId, setEnvStudyId] = useState("");
```
Comment (broarr) [inline src/App.js:17]: Similarly here, I don't think this needs to be state. Is there are reason why you've made this state? Will it change throughout the run of the experiment?

File: src/App.js [hunk 31]
```diff
@@ -1,79 +1,104 @@
...
-                psiturk.completeHIT()
-              }
-              completePsiturk()
-            }
-          },
-        }}
+  return (
+    <>
+      {loggedIn ? (
+        <JsPsychExperiment ipcRenderer={ipcRenderer} psiturk={psiturk} />
```
Comment (broarr) [inline src/App.js:86]: I think it's cleaner to pass a function instead of the `ipcRenderer` object. The `ipcRenderer` object should be used as little as possible as it allows for low-level access to everything in electron. It's so dangerous that in later versions of Electron they disable it by default.

Perhaps send a function as a prop to `JsPsychExperiment` that either sends data via `ipcRenderer` or records data via `psiturk`. You may need two functions there, one for data update and one for finish

File: src/components/JsPsychExperiment.js [hunk 32]
```diff
@@ -0,0 +1,60 @@
...
+              }
+              // Electron case
+              else if ( ipcRenderer ) {
+                ipcRenderer.send('data', data)
+              }
+              // Mturk case
+              else if (psiturk) {
+                  psiturk.recordTrialData(data)
+              }
+            },
```
Comment (broarr) [inline src/components/JsPsychExperiment.js:38]: I'd make this function a prop of `JsPsychExperiment`

File: src/components/JsPsychExperiment.js [hunk 33]
```diff
@@ -0,0 +1,60 @@
...
+              else if (psiturk) {
+                const completePsiturk = async () => {
+                  psiturk.saveData()
+                  await sleep(5000)
+                  psiturk.completeHIT()
+                }
+                completePsiturk()
+              }
+            },
+          }}
```
Comment (broarr) [inline src/components/JsPsychExperiment.js:54]: Same with this function

**Review by broarr — CHANGES_REQUESTED** [review]
Looking great! I think passing `on_data` and `on_finish` to the `JsPsychExperiment` will significantly clean things up, and then I'll approve!

File: src/App.js (hunk 31 above)
Comment (suivo2001) [reply src/App.js:86]: I see. This is what was there in the original App.js so I didn't want to change the things that worked. I'll make these changes and make it similar to the validation function and make sure eveyrthing works.

**Review by broarr — APPROVED** [review]

File: src/lib/utils.js [hunk 34]
```diff
@@ -72,15 +72,10 @@ const images = importAll(requireContext('../assets/images', false, /\.(png|jpe?g
 const getTurkUniqueId = () => {
   const turkInfo = jsPsych.turk.turkInfo()
   const uniqueId = `${turkInfo.workerId}:${turkInfo.assignmentId}`
+  jsPsych.data.addProperties({turk_unique_id: uniqueId})
```
Comment (mcmcgrath13) [inline src/lib/utils.js:75]: the `turk_unique_id` is the participant id right? this should then be added as such

**Review by mcmcgrath13 — CHANGES_REQUESTED** [review]

File: public/electron.js [hunk 35]
```diff
@@ -201,7 +205,7 @@ ipc.on('data', (event, args) => {
     //write the data
     stream.write(JSON.stringify({...args, git}))
 
-    // Copy provocation images to patient's data folder
+    // Copy provocation images to participant's data folder
```
Comment (mcmcgrath13) [inline public/electron.js:208]: this is unrelated to your changes, but line 209 and this comment make no sense for this task - can you delete them?

File: src/App.js [hunk 36]
```diff
@@ -1,79 +1,154 @@
...
-            }
-          },
-        }}
+    // If on desktop
+    if (IS_ELECTRON) {
+      const electron = window.require("electron");
+      const renderer = electron.ipcRenderer;
+      setRenderer(renderer);
+      // If at home, fill in fields based on environment variables
+      if (AT_HOME) {
```
Comment (mcmcgrath13) [inline src/App.js:68]: I don't think this should be specific to at_home, as in the clinic the study ID could definitely be set this way as well.

Comment (mcmcgrath13) [reply src/App.js:68]: then you wouldn't need the else on 78

File: src/components/JsPsychExperiment.js [hunk 37]
```diff
@@ -0,0 +1,25 @@
+import React from "react";
+import { Experiment, jsPsych } from "jspsych-react";
+import { tl } from "../timelines/main";
+import "../App.css";
+import "bootstrap/dist/css/bootstrap.css";
+import "@fortawesome/fontawesome-free/css/all.css";
+import { rt_categorize_html } from "../lib/rt-categorize-html";
+
+function JsPsychExperiment({ dataUpdateFunction, dataFinishFunction }) {
+  console.log("Outside Turk:", jsPsych.turk.turkInfo().outsideTurk);
```
Comment (mcmcgrath13) [inline src/components/JsPsychExperiment.js:10]: remove console logging, should the css imports be at the app level? otherwise, the component here looks great!

File: src/components/Login.js [hunk 38]
```diff
@@ -0,0 +1,86 @@
+import React, { useState, useEffect } from "react";
+import Form from "react-bootstrap/Form";
+import Button from "react-bootstrap/Button";
+import { jsPsych } from "jspsych-react";
+import "../App.css";
+import "bootstrap/dist/css/bootstrap.css";
+import "@fortawesome/fontawesome-free/css/all.css";
```
Comment (mcmcgrath13) [inline src/components/Login.js:7]: same question on css imports - just at app level should work

File: src/components/Login.js [hunk 39]
```diff
@@ -0,0 +1,86 @@
...
+    return participantId.length > 0 && studyId.length > 0;
+  }
+
+  // Function to log in participant
+  function handleSubmit(e) {
+    e.preventDefault();
+    // Validates fields
+    validationFunction(participantId, studyId, startDate, timestamp)
+    // Logs in depending on result from promise
+    .then((loggedIn) => {
```
Comment (mcmcgrath13) [inline src/components/Login.js:35]: I still think the jspsych piece of this should move to app.  create a wrapper function around `setLogin`, which takes the study and participant id, then if the login state is true, adds them to jspsych.  This would also be able to be used by the mturk piece of code.

It would look something like (in app.js)
```js
const setLoggedIn = (loggedIn, study_id, participant_id) => {
  if (loggedIn) {
    jsPsych.data.addProperties(...)
  }
  setLogin(loggedIn)
}
```

File: src/trials/userId.js [hunk 40]
```diff
@@ -1,7 +1,10 @@
 import { lang, MTURK } from '../config/main'
-import { getUserId, getTurkUniqueId } from '../lib/utils'
+import { getTurkUniqueId } from '../lib/utils'
 import { baseStimulus } from '../lib/markup/stimuli'
 
+
```
Comment (mcmcgrath13) [inline src/trials/userId.js:5]: can this file be deleted now?

**Review by mcmcgrath13 — CHANGES_REQUESTED** [review]

File: src/App.js [hunk 41]
```diff
@@ -51,6 +57,20 @@ function App() {
     completePsiturk();
   };
 
+  // Function to add jspsych data on login
+  const setLoggedIn = useCallback(
```
Comment (mcmcgrath13) [inline src/App.js:61]: what's the reasoning around `useCallback` vs a plain function? (just curious, I'm not familiar with it)

File: src/App.js [hunk 42]
```diff
@@ -88,15 +103,8 @@ function App() {
...
-        // study id yet to be determined
-        jsPsych.data.addProperties({
-          participant_id: turkId,
-          study_id: "mturk",
-          timestamp: dateTimestamp.toString(),
-          start_date: curDate.toString(),
-        });
         setMethod("mturk");
-        setLogin(true);
+        setLoggedIn(true, "mturk", turkId, startDate, timestamp)
```
Comment (mcmcgrath13) [inline src/App.js:107]: why do the startDate and timestamp need to be passed if they're part of the state of the callback? (also, just curious)

File: src/App.js [hunk 43]
```diff
@@ -1,79 +1,162 @@
...
+      }
+      setMethod("desktop");
+      // If online
+    } else {
+      // If MTURK
+      if (MTURK) {
+        /* eslint-disable */
+        window.lodash = _.noConflict();
+        const turkId = getTurkUniqueId();
+        const dateTimestamp = Date.now();
```
Comment (mcmcgrath13) [inline src/App.js:103]: are these timestamps used?

**Review by mcmcgrath13 — CHANGES_REQUESTED** [review]

**Review by mcmcgrath13 — APPROVED** [review]
🚀

**Review by mirestrepo — APPROVED** [review]
